### PR TITLE
[Coop] Mostly convert System.GC and System.Runtime.InteropServices.GC.Handle.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -8425,3 +8425,26 @@ mono_register_jit_icall (gconstpointer func, const char *name, MonoMethodSignatu
 	return mono_register_jit_icall_full (func, name, sig, no_wrapper, NULL);
 }
 
+ICALL_EXPORT int
```
Comment (jaykrell) [inline mono/metadata/icall.c]: One wonders if this can overflow.

Comment (jaykrell) [reply mono/metadata/icall.c]: It follows MSDN though https://msdn.microsoft.com/en-us/library/system.gc.collectioncount(v=vs.110).aspx.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Familiar Rosyln and FullAOT+LLVM crashes.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 FullAOT+LLVM

File: mono/metadata/gc.c [hunk 2]
```diff
@@ -622,70 +629,73 @@ ves_icall_System_GC_WaitForPendingFinalizers (void)
...
+ves_icall_System_GC_register_ephemeron_array (MonoObjectHandle array, MonoError *error)
 {
-#ifdef HAVE_SGEN_GC
-	if (!mono_gc_ephemeron_array_add (array)) {
-		mono_set_pending_exception (mono_object_domain (array)->out_of_memory_ex);
-		return;
-	}
-#endif
+	if (!mono_gc_ephemeron_array_add (MONO_HANDLE_RAW (array)))
+		mono_set_pending_exception (MONO_HANDLE_DOMAIN (array)->out_of_memory_ex);
```
Comment (lambdageek) [inline mono/metadata/gc.c]: Could drop the use of `mono_set_pending_exception` in favor of `mono_error_set_exception_instance (error, ...); return;`

File: mono/metadata/gc.c [hunk 3]
```diff
@@ -622,70 +629,73 @@ ves_icall_System_GC_WaitForPendingFinalizers (void)
...
-ves_icall_System_GC_register_ephemeron_array (MonoObject *array)
+ves_icall_System_GC_register_ephemeron_array (MonoObjectHandle array, MonoError *error)
 {
-#ifdef HAVE_SGEN_GC
-	if (!mono_gc_ephemeron_array_add (array)) {
-		mono_set_pending_exception (mono_object_domain (array)->out_of_memory_ex);
-		return;
-	}
-#endif
+	if (!mono_gc_ephemeron_array_add (MONO_HANDLE_RAW (array)))
```
Comment (lambdageek) [inline mono/metadata/gc.c]: Why not just keep the #ifdef here instead of adding a stub to Boehm?

File: mono/metadata/object-internals.h [hunk 4]
```diff
@@ -46,7 +46,7 @@
 
 /* Use this as MONO_CHECK_ARG_NULL_HANDLE (arg,) in functions returning void */
 #define MONO_CHECK_ARG_NULL_HANDLE(arg, retval) do { 		\
-	if (MONO_HANDLE_IS_NULL (arg))				\
+	if (G_UNLIKELY (MONO_HANDLE_IS_NULL (arg)))		\
```
Comment (jaykrell) [inline mono/metadata/object-internals.h]: "likely" has bounced here, somewhat by accident.
Arguably it should  be in MONO_HANDLE_IS_NULL.

File: mono/metadata/object-internals.h [hunk 5]
```diff
@@ -55,7 +55,7 @@
 
 /* Use this as MONO_CHECK_NULL (arg,) in functions returning void */
 #define MONO_CHECK_NULL(arg, retval) do { 			\
```
Comment (jaykrell) [inline mono/metadata/object-internals.h]: This macro only has one user at this point.

File: mono/metadata/object-internals.h (hunk 4 above)
Comment (lambdageek) [reply mono/metadata/object-internals.h]: I don't think it makes sense for `likely`/`unlikey` to be part of `MONO_HANDLE_IS_NULL` - different circumstances would have different likelihoods

File: mono/metadata/gc.c (hunk 3 above)
Comment (jaykrell) [reply mono/metadata/gc.c]: ifdef is bad and to be reduced. Variation generally better at link time or runtime (command line switch), in order to compile and/or link fewer times. Unless for extreme perf reasons.

File: mono/metadata/object-internals.h (hunk 4 above)
Comment (jaykrell) [reply mono/metadata/object-internals.h]: Agreed, though barely. Pretty few paths will have l likely(null), I think.  But I understand, everything is context relative.

File: mono/metadata/gc.c (hunk 2 above)
Comment (jaykrell) [reply mono/metadata/gc.c]: Good point, I'll try.

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Pull request merged**